src/libraries/System.Net.Quic/src/System/Net/Quic/QuicConnection.cs

```diff
@@ -69,14 +69,51 @@ static async ValueTask<QuicConnection> StartConnectAsync(QuicClientConnectionOpt
 
             using CancellationTokenSource linkedCts = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
 
-            if (options.HandshakeTimeout != Timeout.InfiniteTimeSpan && options.HandshakeTimeout != TimeSpan.Zero)
-            {
-                linkedCts.CancelAfter(options.HandshakeTimeout);
-            }
-
             try
             {
-                await connection.FinishConnectAsync(options, linkedCts.Token).ConfigureAwait(false);
+                if (!options.RemoteEndPoint.TryParse(out string? host, out IPAddress? address, out int port))
+                {
+                    throw new ArgumentException(SR.Format(SR.net_quic_unsupported_endpoint_type, options.RemoteEndPoint.GetType()), nameof(options));
+                }
+
+                if (address is not null && host is null)
+                {
+                    if (options.HandshakeTimeout != Timeout.InfiniteTimeSpan && options.HandshakeTimeout != TimeSpan.Zero)
+                    {
+                        linkedCts.CancelAfter(options.HandshakeTimeout);
+                    }
+
+                    await connection.FinishConnectAsync(options, null, address, port, linkedCts.Token).ConfigureAwait(false);
+                    return connection;
+                }
+
+                Debug.Assert(host is not null);
+                IPAddress[] addresses = await Dns.GetHostAddressesAsync(host, cancellationToken).ConfigureAwait(false);
+                cancellationToken.ThrowIfCancellationRequested();
+                if (addresses.Length == 0)
+                {
+                    throw new SocketException((int)SocketError.HostNotFound);
+                }
+
+                ExceptionDispatchInfo? lastException = null;
+
+                foreach (IPAddress addr in addresses)
+                {
+                    connection = new QuicConnection();
```

Comment:
This throws away the first QuicConnection instance allocated at the beginning of the function